c_reference/tests/kws/test_keyword_spotting.c

```diff
@@ -0,0 +1,272 @@
+  };
+
+  ConvLayers_LR_Params point_param_5 = {
+    .W1 = CNN5_POINT_W1,
+    .W2 = CNN5_POINT_W2,
+    .B = CNN5_POINT_BIAS,
+    .rank = POST_CNN_LOW_RANK
+  };
+
+  FastGRNN_LR_Params bwd_RNN_params = {
+    .mean   = 0,
+    .stdDev = 0,
+    .W1     = B_W1,
+    .W2     = B_W2,
+    .wRank  = RNN_LOW_RANK,
+    .U1     = B_U1,
+    .U2     = B_U2,
+    .uRank  = RNN_LOW_RANK,
+    .Bg     = B_BIAS_GATE,
+    .Bh     = B_BIAS_UPDATE,
+    .sigmoid_zeta = sigmoid(B_ZETA),
+    .sigmoid_nu   = sigmoid(B_NU)
+  };
+
+  FastGRNN_LR_Params fwd_RNN_params = {
+    .mean   = 0,
+    .stdDev = 0,
+    .W1     = F_W1,
+    .W2     = F_W2,
+    .wRank  = RNN_LOW_RANK,
+    .U1     = F_U1,
+    .U2     = F_U2,
+    .uRank  = RNN_LOW_RANK,
+    .Bg     = F_BIAS_GATE,
+    .Bh     = F_BIAS_UPDATE,
+    .sigmoid_zeta = sigmoid(F_ZETA),
+    .sigmoid_nu   = sigmoid(F_NU)
+  };
+
+  float preComp[RNN_I_F] = { 0.0 };
+  float tempLRW[RNN_LOW_RANK] = { 0.0 };
+  float tempLRU[RNN_LOW_RANK] = { 0.0 };
+  float normFeatures[RNN_I_F] = { 0.0 };
+  FastGRNN_LR_Buffers buffers = {
+    .preComp = preComp,
+    .tempLRW = tempLRW,
+    .tempLRU = tempLRU,
+    .normFeatures = normFeatures
+  };
+  
+  unsigned in_T, out_T;
+
+  /* Pre-CNN */
+  in_T = I_T;
+  out_T = in_T - PRE_CNN_FILT + (PRE_CNN_FILT_PAD<<1) + 1; // Depth pad = 2 for kernel size =5. SAME PAD
+  float* cnn1_out = (float*)malloc(out_T * PRE_CNN_O_F * sizeof(float));
+  // Since bnorm is the first layer and inplce will alter input. Use only if input can be discarded/altered. Else avoid inplace
+  dscnn_lr(cnn1_out, mem_buf, in_T, PRE_CNN_I_F,
+    BNORM_CNN1_MEAN, BNORM_CNN1_VAR, 0, 0, 0, PRE_CNN_BNORM_INPLACE,
+    PRE_CNN_O_F, PRE_CNN_FILT_PAD, PRE_CNN_FILT,
+    &conv_params, PRE_CNN_FILT_ACT); // regular tanh activation
+
+  batchnorm1d(0, cnn1_out, in_T, RNN_I_F, 
+    BNORM_RNN_MEAN, BNORM_RNN_VAR, 0, 0, 0, 1, 0.00001); // Currently In-place only
+
+  /* Bricked Bi-FastGRNN Block */
+  int rnn_hidden = RNN_O_F>>1, out_index = 0;
+
+  out_T = in_T/HOP + 1;
+  float* temp_hiddenstate = (float*)calloc(rnn_hidden, sizeof(float));
+  float* rnn_out = (float*)malloc(out_T * RNN_O_F * sizeof(float));
+  // Forward Pass
+  for (int t = 0; t < FWD_WINDOW; t++) {
+    fastgrnn_lr(temp_hiddenstate, rnn_hidden,
+      cnn1_out + (t * RNN_I_F) , RNN_I_F, 1,
+      &fwd_RNN_params, &buffers, 0, 0);
+    if (t % HOP==0)
+      memcpy(rnn_out + ((out_index++)*RNN_O_F), temp_hiddenstate, rnn_hidden*sizeof(float));
+  }
+  memcpy(rnn_out + ((out_index++)*RNN_O_F), temp_hiddenstate, rnn_hidden*sizeof(float));
+  for (int t = HOP; t <= in_T - FWD_WINDOW; t += HOP ) {
+    memset(temp_hiddenstate, 0, rnn_hidden*sizeof(float));
+    fastgrnn_lr(temp_hiddenstate, rnn_hidden,
+      cnn1_out + (t * RNN_I_F) , RNN_I_F, FWD_WINDOW,
+      &fwd_RNN_params, &buffers, 0, 0);
+    memcpy(rnn_out + ((out_index++)*RNN_O_F), temp_hiddenstate, rnn_hidden*sizeof(float));    
+  }
+
+  // Backward Pass
```

Comment:
forward and backward passes could be explained a bit in text here so the reader can follow the intent.

Reply:
I've moved the bricking to a separate file altogether. This bricking module has been added to the src folder, so that it can be used as function for other applications as well.

I've added a detailed documentation and line comments explaining how the code works. I've even included notes on how to use the bi-directional option and have highlighted the constrains for the bi-directional application